src/main/java/org/terracotta/offheapstore/paging/UpfrontAllocatingPageSource.java

```diff
@@ -440,67 +473,146 @@ public synchronized Runnable removeAllocationThreshold(ThresholdDirection direct
-            currentChunkSize >>>= 1;
-            lastFailureAt = allocated;
-          }
-        } else {
-          buffers.add(b);
-          allocated += b.capacity();
-          
-          if (allocatorLog != null) {
-            allocatorLog.printf("%d,%d,%d,%d,%d,%d,%d,%d%n", System.nanoTime() - start, blockDuration, b.capacity(), allocated, PhysicalMemory.freePhysicalMemory(), PhysicalMemory.totalSwapSpace(), PhysicalMemory.freeSwapSpace(), PhysicalMemory.ourCommittedVirtualMemory());
-          }
-          if (allocated > nextProgressLogAt) {
-            LOGGER.info("Allocation {}% complete", (100 * allocated) / max);
-            nextProgressLogAt += progressStep;
+        if (allocatorLog != null) {
+          allocatorLog.printf("timestamp,duration,size,physfree,totalswap,freeswap,committed%n");
+        }
+
+        ExecutorService executorService = Executors.newFixedThreadPool(Runtime.getRuntime().availableProcessors());
+
+        List<Future<Long>> futures = new ArrayList<Future<Long>>((int)toAllocate / maxChunk + 1);
+
+        for (long dispatched = 0; dispatched < toAllocate; ) {
+          final int currentChunkSize = (int)Math.min(maxChunk, toAllocate - dispatched);
+          futures.add(executorService.submit(new Callable<Long>() {
+            @Override
+            public Long call() throws Exception {
+              return bufferAllocation(source, currentChunkSize, minChunk, fixed, allocatorLog, start, buffers);
+            }
+          }));
+          dispatched += currentChunkSize;
+        }
+
+        executorService.shutdown();
+
+        long allocated = 0;
+        long progressStep = Math.max(PROGRESS_LOGGING_THRESHOLD, (long)(toAllocate * PROGRESS_LOGGING_STEP_SIZE));
+        long nextProgressLogAt = progressStep;
+
+        for (Future<Long> future : futures) {
+          try {
+            allocated += future.get();
+            if (allocated > nextProgressLogAt) {
+              LOGGER.info("Allocation {}% complete", (100 * allocated) / toAllocate);
+              nextProgressLogAt += progressStep;
+            }
+          } catch (ExecutionException e) {
+            if (e.getCause() instanceof RuntimeException) {
+              throw (RuntimeException)e.getCause();
+            }
+            throw new RuntimeException(e);
+          } catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
+            break;
           }
         }
+      } finally {
+        if (allocatorLog != null) {
+          allocatorLog.close();
+        }
       }
-      if (allocatorLog != null) {
-        allocatorLog.close();
+
+      if(LOGGER.isInfoEnabled()) {
+        long duration = System.nanoTime() - start;
+        LOGGER.info("Took {} ms to create off-heap storage of {}B.", TimeUnit.NANOSECONDS.toMillis(duration), DebuggingUtils.toBase2SuffixedString(toAllocate));
```

Comment:
Because since we are logging the Allocation completion percentage in info, it seemed meaningful to also give the final result in info. No?